libflux/core/src/scanner/mod.rs

```diff
@@ -1,43 +1,47 @@
 #![allow(missing_docs)]
-include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
 
 pub type CChar = u8;
 
 use crate::fmt;
 use std::collections::HashMap;
 use std::ffi::CString;
 use std::str;
+use std::vec::Vec;
 
-pub mod rust;
-use rust::Scan;
+#[rustfmt::skip]
+#[allow(clippy::all)]
+mod scanner;
+
+mod token;
+pub use token::TokenType;
 
 pub struct Scanner {
-    data: CString,
-    ps: *const CChar,
-    p: *const CChar,
-    pe: *const CChar,
-    eof: *const CChar,
-    last_newline: *const CChar,
-    cur_line: u32,
-    checkpoint: *const CChar,
-    checkpoint_line: u32,
-    checkpoint_last_newline: *const CChar,
-    token: TOK,
+    data: Vec<u8>,
+    ps: i32,
```

Comment:
This makes me raise an eyebrow, as a `CChar` is a `u8`, so we're taking what was a single byte and making it a signed 32-bit integer. I realize that `scan` has different ideas on its interface, but maybe that interface should change.

I'd prefer continuing to use a `u8` if we can get away with it (not sure if we need the type alias). Thoughts?

Reply:
When you switch from scanning in C to scanning in Rust, some of the variables you need to populate and make available to the ragel-generated must change in type. C scans using `p` and `pe` as pointers to the input. In rust you define the input array `data`, and then `p` and `pe` become integers indexing the `data` array.

Also to note, the type of the `data` array must correspond to the type of the alphabet defined in the ragel file (see `alphtype`). That ensures that ragel and the host code are in agreement about how to interpret the input data. As things are in the ragel grammar, we must use `u8`, as all the literal tokens are defined in terms of `u8`.

Reply:
@adrian-thurston could we use `&[u8]` as data, rather than a `Vec`? It's essentially immutable, so doesn't need the functionality of `Vec`, and would probably benefit from _requiring_ sequential access, particularly for CPU cache use. What about `&[char]` even, which is equivalent, but carries a semantic value of saying "this should actually be a character we read from the string"?

Reply:
Please ignore the comment about `char` - rust's `char` is not actually equivalent to `u8` because `char` can be any unicode character (which means more than just one byte).

Reply:
Could use `&[u8]` but not `&[char]` as they won't be interpreted right, because all the ragel code specifies patterns using unsigned single bytes notation (in ragel syntax). If they were all specified as signed, four-byte scalars in ragel, we could use `char`.

My rust sucks now that I've been in go so long ... but I can't seem to make `&[u8]` work because of borrowing, I think we need to copy it in using `[u8]` because the scanner outlives the data that's passed in. But hopefully the copy-in should be optimized away?

Reply:
```
19 |     data: [u8],
   |           ^^^^ doesn't have a size known at compile-time
```
Maybe this is why I used a vec? It's either `Vec<u8>` or a boxed `&[u8]`?

Reply:
Again, my rust is not super-good right now, but I can only make `Vec<u8>` work, can't copy it into any combination of `Box`, `&`, or `[u8]`.  All result in unfixable compile errors for me.

Reply:
Yeah, there's likely some explicit lifetime work that we'll need to do in order to make it work. It's probably premature optimization at this point.